### ISSUE TITLE
travis-ci: migrate libbpf CI from Travis CI to Github Actions

### ISSUE THREAD

**Issue by anakryiko** [libbpf/libbpf#316]
We need to migrate libbpf CI from the current setup with Travis CI to something like Github Actions to avoid paying for Travis CI ($69 per month) while getting only single build at a time. From the looks of it, Github Actions is both free and would provide better and faster CI capabilities.

https://github.com/libbpf/libbpf/blob/master/.travis.yml is the starting point.

**Comment by tohojo** [comment]
We have the same issue for xdp-tools (cf https://github.com/xdp-project/xdp-tools/issues/114) - linking the issues to subscribe to any updates here :)

**Comment by tsipa** [comment]
https://github.com/tsipa/vmtest/commit/d170ddc185e539f40855a7838916d4775708129b i did necessary changes to make syntax compatible with gh actions, but gh will not let us run qemu  inside their containers.

The explanation is available here https://github.com/actions/virtual-environments/issues/183
TL;DR "azure something-something, github something-something, we won't support it in a human life".


I don't understand the subject area to justify if we can run these tests inside the docker container wo VM, but if no - we have to consider something else, not GH actions.

**Comment by tohojo** [comment]
I did some experiments with getting Qemu to run in software-emul mode here: https://github.com/xdp-project/xdp-tools/pull/113

It runs on Travis, but haven't tried converting it to GH actions; care to try that out? It's just adding a couple of qemu args (`-machine accel=tcg -cpu qemu64`) so should be simple enough to adapt, you'll just need to get rid of the kvm stuff of course...

**Comment by tsipa** [comment]
Thanks for this work, will try be eod as i already have dev env set up.

But i strongly suspect without hardware virtualization it's going to be a torture and we likely to hit many weird issues caused by inadequate timeouts here and there - SW emulation is really slow.

**Comment by tohojo** [comment]
"Sergey V. Yudin" ***@***.***> writes:

> Thanks for this work, will try be eod as i already have dev env set
> up.

Awesome!

> But i strongly suspect without hardware virtualization it's going to
> be a torture and we likely to hit many weird issues caused by
> inadequate timeouts here and there - SW emulation is really slow.

Yeah, but maybe that'll be an excuse to rely less on timeouts, and thus
get more reliable tests? One can hope, at least ;)

**Comment by tsipa** [comment]
Wow, it worked and it wasn't that bad:
travis 32min
https://travis-ci.com/github/kernel-patches/bpf/builds/228452123
GH actions 36min
https://github.com/tsipa/bpf-next/runs/2798310036

https://github.com/tsipa/vmtest/commit/ecf3a9b872bb34293f2d0fbe9855d76927df419f i tried this.

I'll prepare the commit and will make PR today or tomorrow.

**Comment by tohojo** [comment]
"Sergey V. Yudin" ***@***.***> writes:

> Wow, it worked and it wasn't that bad:
> travis 32min
> https://travis-ci.com/github/kernel-patches/bpf/builds/228452123
> GH actions 36min
> https://github.com/tsipa/bpf-next/runs/2798310036
>
> https://github.com/tsipa/vmtest/commit/ecf3a9b872bb34293f2d0fbe9855d76927df419f i tried this.
>
> I'll prepare the commit and will make PR today or tomorrow.

Awesome! Thanks for looking at this :)

**Comment by yonghong-song** [comment]
cc @davemarchevsky for an example of github actions

**Comment by anakryiko** [comment]
> Wow, it worked and it wasn't that bad:
> travis 32min
> https://travis-ci.com/github/kernel-patches/bpf/builds/228452123
> GH actions 36min
> https://github.com/tsipa/bpf-next/runs/2798310036

Great!

My biggest hope is that we'll get back the parallelism of running multiple build steps for libbpf CI, so +4 minutes doesn't matter much to me.

> 
> [tsipa/vmtest@ecf3a9b](https://github.com/tsipa/vmtest/commit/ecf3a9b872bb34293f2d0fbe9855d76927df419f) i tried this.
> 
> I'll prepare the commit and will make PR today or tomorrow.

That's great, looking forward to finally be able to jump out of Travis CI.


> But i strongly suspect without hardware virtualization it's going to be a torture and we likely to hit many weird issues caused by inadequate timeouts here and there - SW emulation is really slow.

We'll fix tests that rely on timeout too much or have too agressive timeouts.

Right now it's actually some perf_branch-related tests that needs fixing, it seems. @danobi  can you please take a look if we can add some skipping logic if hardware is missing the LBR feature? Or if it's already there, why it doesn't kick in?

**Comment by danobi** [comment]
@anakryiko where is the error for LBR feature?

On https://github.com/tsipa/bpf-next/runs/2798310036 I see:

> 2021-06-10T23:45:31.3861987Z Summary: 159/911 PASSED, 6 SKIPPED, 2 FAILED

> 2021-06-10T23:42:28.7505839Z ./test_progs:udp_redir_to_connected:1691: egress: read: Resource temporarily unavailable
> 2021-06-10T23:42:28.7506798Z udp_redir_to_connected:FAIL:1691
> 2021-06-10T23:42:28.7513933Z #112/36 sockmap IPv4 test_udp_redir:FAIL

> 2021-06-10T23:42:54.3159485Z #112 sockmap_listen:FAIL

**Comment by anakryiko** [comment]
@danobi it's on Travis CI run (https://travis-ci.com/github/kernel-patches/bpf/builds/228452123):

```
#76/1 perf_branches_hw:OK
test_perf_branches_no_hw:PASS:perf_event_open 0 nsec
test_perf_branches_common:PASS:test_perf_branches_load 0 nsec
test_perf_branches_common:PASS:attach_perf_event 0 nsec
test_perf_branches_common:PASS:set_affinity 0 nsec
check_bad_sample:FAIL:output not valid no valid sample from prog
#76/2 perf_branches_no_hw:FAIL
#76 perf_branches:FAIL
```

**Comment by danobi** [comment]
@anakryiko ah I see. I wonder if it's b/c of https://github.com/torvalds/linux/blob/6b00bc639f1f2beeff3595e1bab9faaa51d23b01/tools/testing/selftests/bpf/progs/test_perf_branches.c#L38-L40

What do you think about a change like: http://ix.io/3qar

**Comment by anakryiko** [comment]
So basically add few retries? Ok, let's try that.

**Comment by danobi** [comment]
Yeah, add retries to work around the fact that LBRs may not be collected.

How do I test this? Other than checking it in upstream

**Comment by anakryiko** [comment]
Try locally. There is no way to test kernel changes with CI without committing them upstream.

**Comment by tsipa** [comment]
Sounds like a legit use-case, we may implement this if needed. Not today or tomorrow, but i'd estimate it as a med/small size item.

**Comment by danobi** [comment]
> Try locally. There is no way to test kernel changes with CI without committing them upstream.

The diff I linked definitely works :) . I tried it before commenting here.

**Comment by anakryiko** [comment]
Done.
